numpy/core/src/multiarray/array_coercion.c

```diff
@@ -0,0 +1,1437 @@
+ * When creating an array with ``np.array(...)`` we have to do two main things:
+ *
+ * 1. Find the exact shape of the resulting array
+ * 2. Find the correct dtype of the resulting array.
+ *
+ * In most cases these two things are can be done in a single processing step.
+ * There are in principle three different calls that should be distinguished:
+ *
+ * 1. The user calls ``np.array(..., dtype=np.dtype("<f8"))``
+ * 2. The user calls ``np.array(..., dtype="S")``
+ * 3. The user calls ``np.array(...)``
+ *
+ * In the first case, in principle only the shape needs to be found. In the
+ * second case, the DType class (e.g. string) is already known but the DType
+ * instance (e.g. length of the string) has to be found.
+ * In the last case the DType class needs to be found as well. Note that
+ * it is not necessary to find the DType class of the entire array, but
+ * the DType class needs to be found for each element before the actual
+ * dtype instance can be found.
+ *
+ * Further, there are a few other things to keep in mind when coercing arrays:
+ *
+ *   * For UFunc promotion, Python scalars need to be handled specially to
+ *     allow value based casting.  This requires python complex/float to
+ *     have their own DTypes.
+ *   * It is necessary to decide whether or not a sequence is an element.
+ *     For example tuples are considered elements for structured dtypes, but
+ *     otherwise are considered sequences.
+ *     This means that if a dtype is given (either as a class or instance),
+ *     it can effect the dimension discovery part.
+ *     For the "special" NumPy types structured void and "c" (single character)
+ *     this is special cased.  For future user-types, this is currently
+ *     handled by providing calling an `is_known_scalar` method.  This method
+ *     currently ensures that Python numerical types are handled quickly.
+ *
+ * In the initial version of this implementation, it is assumed that dtype
+ * discovery can be implemented sufficiently fast.  That is, it is not
+ * necessary to create fast paths that only find the correct shape e.g. when
+ * ``dtype=np.dtype("f8")`` is given.
+ *
+ * The code here avoid multiple conversion of array-like objects (including
+ * sequences). These objects are cached after conversion, which will require
+ * additional memory, but can drastically speed up coercion from from array
+ * like objects.
+ */
+
+
+/*
+ * For finding a DType quickly from a type, it is easiest to have a
+ * a mapping of pytype -> dtype.
+ * Since a DType must know its type, but the type not the DType, we will
+ * store the DType as a weak reference. When a reference is dead we can
+ * remove the item from the dictionary.
+ * A cleanup should probably be done occasionally if (and only if) a large
+ * number of type -> DType mappings are added.
+ * This assumes that the mapping is a bifurcation DType <-> type
+ * (there is exactly one DType for each type and vise versa).
```

Comment:
Hmmmpf, I have to think about it more.  Yes, you are probably right, and I am now not sure the logic is correct to begin with.  We need _some_ mapping though (unless we modify the python types). In this case, I think I actually got it the wrong way around... If the python type is weakly referenced in the dictionary, once its deleted we can remove the entry from the mapping.  That implies the `DType` must have been deleted before (since it must reference the type), although that could be part of a cycle resolution.

I want to move this forward, and this is all about deleting DTypes.  Would we be fine with assuming that there is a reasonable solution to this (maybe not with a dictionary mapping), and for now implementing it without support for deleting DTypes or scalar types? I.e. true support for dynamic type<->DType pair creations?

Reply:
Deleted it with a comment, will push soon)